casr/src/triage.rs

```diff
@@ -0,0 +1,464 @@
+        {
+            info!("There are less than 2 CASR reports, nothing to cluster.");
+            return summarize_results(matches, crashes);
+        }
+        info!("Clustering CASR reports...");
+        let casr_cluster_c = Command::new(&casr_cluster)
+            .arg("-c")
+            .arg(output_dir.clone().into_os_string())
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cluster:?}"))?;
+
+        if casr_cluster_c.status.success() {
+            info!(
+                "{}",
+                String::from_utf8_lossy(&casr_cluster_c.stdout).trim_end()
+            );
+        } else {
+            error!("{}", String::from_utf8_lossy(&casr_cluster_c.stderr));
+        }
+
+        // Remove reports from deduplication phase. They are in clusters now.
+        for casrep in fs::read_dir(output_dir)?.flatten().map(|e| e.path()) {
+            if let Some(ext) = casrep.extension() {
+                if ext == "casrep" {
+                    let _ = fs::remove_file(casrep);
+                }
+            }
+        }
+    }
+
+    summarize_results(matches, crashes)
+}
+
+/// Copy crashes next to reports and print summary.
+/// Run casr-gdb on uninstrumented binary if specified in ARGS.
+/// Print analysis statistic.
+///
+/// # Arguments
+///
+/// * `matches` - tool arguments
+///
+/// * `crashes` - set of crashes, specified as a CrashInfo structure
+fn summarize_results(
+    matches: &clap::ArgMatches,
+    crashes: &HashMap<String, CrashInfo>,
+) -> Result<()> {
+    // Get output dir
+    let dir = matches.get_one::<PathBuf>("output").unwrap();
+    // Copy crashes next to reports
+    copy_crashes(dir, crashes)?;
+
+    // Get casr-gdb args
+    let tool = std::env::current_exe()?
+        .file_name()
+        .unwrap()
+        .to_str()
+        .unwrap()
+        .to_string();
+    let gdb_args = if tool.eq("casr-libfuzzer") {
+        // casr-libfuzzer case
+        if let Some(argv) = matches.get_one::<String>("casr-gdb-args") {
+            shell_words::split(argv)?
+        } else {
+            Vec::new()
+        }
+    } else if let Some(argv) = matches.get_many::<String>("ARGS") {
+        // casr-afl case
+        argv.cloned().collect()
+    } else {
+        Vec::new()
+    };
+    // Get timeout
+    let timeout = if let Some(timeout) = matches.get_one::<u64>("timeout") {
+        *timeout
+    } else {
+        0
+    };
+    // Get number of threads
+    let jobs = if let Some(jobs) = matches.get_one::<u32>("jobs") {
+        *jobs as usize
+    } else {
+        std::cmp::max(1, num_cpus::get() / 2)
+    };
+
+    if !gdb_args.is_empty() {
+        let casr_gdb = get_path("casr-gdb")?;
+        // Run casr-gdb on uninstrumented binary.
+        let crashes: Vec<_> = WalkDir::new(dir)
+            .into_iter()
+            .filter_entry(|e| {
+                let name = e.file_name().to_str().unwrap();
+                !name.eq("clerr") && !name.eq("oom") && !name.eq("timeout")
```

Comment:
My bad. We can still try generating reports for clerr. casr-san may fail and casr-gdb will help with generating reports for such cases.

Reply:
removed check